Create NPM package

icheered opened the pull request:
This project is awesome so I did my best to turn it into an NPM package so that it can be installed easily. A few things to note:

* To publish the package to NPM you only need to run the following commands after cloning the repository: `npm login` and then `npm publish`. Thats it! I thought the original creator should publish the package as you have direct control over the source code.
* I have added documentation in the readme because the readme will also appear on npm. 
* I added the post-build `libapi.*` files (and removed them from the .gitignore) because these files are uploaded to NPM. On subsequent releases you can just rebuild the package, change the version number, and run `npm publish` again.

RReverser commented:
Nice, thanks!

Some thoughts:

 - I think some npm metadata will need to be updated, I'll take care of this.
 - I wonder if we can reuse the version of libapi.mjs that is already built by Github Actions. It's currently "published" to this branch https://github.com/GoogleChromeLabs/web-gphoto2/blob/gh-pages/.github/workflows/main.yml which hosts gh-pages, but perhaps it could also be reused so that build artifact is stored separately from code itself. I think it should be possible to do via this workflow https://docs.github.com/en/actions/publishing-packages/publishing-nodejs-packages - maybe you could add a GH action, and I'll later add a npm token myself.
 - Minor nit but it should be always "Wasm" not "WASM" as it's not abbreviation. Or we can just spell out "WebAssembly" in some places for search discoverability.
 - Maybe it's worth providing a higher-level wrapper that takes care of `connection.schedule` calls? It was fine for the demo app, but I imagine a lot of users might struggle / forget to wrap camera calls into those, but they're important to avoid "USB device busy" errors.

icheered commented:
@RReverser  I've made an attempt at a wrapper class named 'Camera' that automatically manages the `connection.schedule` calls, making it easier for users to interact with the library. The class also incorporates a conditional import from `libapi.mjs`, ensuring that the library is only loaded within a browser environment. This provides the added advantage of saving the user from implementing ~20 lines of extra code themselves.

During my initial tests, I've verified that the following functionalities operate as expected:
- `camera.connect`
- `camera.getSupportedOps`
- `camera.getConfig`
- `camera.setConfigValue`
- `camera.captureImageAsFile`

Although I haven't extensively tested `camera.capturePreviewAsBlob` due to the complexity of rendering the video, initial indications show that the function behaves as intended and returns the appropriate data.

As for automating package releases, I'll certainly explore the possibilities. However, considering that the repository's codebase isn't likely to undergo frequent updates, the necessity for automating the process seems somewhat diminished. Manual publication of a new release is a straightforward command (`npm publish') away. 
Additionally I'm aware that storing build artifacts in the repository isn't a widely favored approach. However, given the relatively small size of these files, I believe the current setup presents an acceptable compromise without requiring elaborate workarounds.

Feel free to make changes to the code in the PR before merging, I'm note particularly well-versed in writing these kinds of wrapper classes and I wouldn't be surprised if I made some mistakes or missed easy opportunities for improvement.

RReverser commented:
> However, considering that the repository's codebase isn't likely to undergo frequent updates, the necessity for automating the process seems somewhat diminished.

Yeah that's fair. Automation could in theory make it easier to update libgphoto2 automatically in the future as well, but for that the changed would need to be backported in a PR first. Maybe something to leave for the future.

If we do keep it in the repo, could you at least put built artifacts into a separate library (e.g. `dist`) and explicitly exclude it from Github diffs via https://docs.github.com/en/repositories/working-with-files/managing-files/customizing-how-changed-files-appear-on-github.

plainsquirrel commented:
정말 감사합니다. 저는 지난 몇주간 이 레포지트리를 Sveltekit, Electron 환경에서 사용하고자 시도하였습니다. 하지만 저의 낮은 수준의 개발실력으로 앞서 언급된 문제를 풀지 못했습니다. 정말 기대가 됩니다.

아주 감사합니다.

plainsquirrel commented:
Thank you very much. I've been trying to use this repository in the Sveltekit, Electron environment for the past few weeks. However, I was unable to solve the aforementioned problem due to my low level of development skills. I'm really looking forward to it.

Thank you very much.

RReverser commented on README.md:
````diff
@@ -1,9 +1,124 @@
+}
+```
+
+A short example on how to use this package to obtain the camera config:
````
This probably can be updated now that you wrapped / automated away those scheduling calls?

icheered replied:
Hmm it seems that the version you were reviewing did not have all the changes because I *did* remove that from the readme on my side. Anyways it should be fixed now, including an updated example.

RReverser commented on README.md:
````diff
@@ -1,9 +1,124 @@
+```ts
+let initModule: any;
+
+if (typeof window !== 'undefined') {
````
I'd probably simplify this to use static import instead. If someone tries to load it in Node.js, the module will error out anyway so no need for separate codepath in the example, better to keep it simple.

RReverser commented on ui/camera.js:
```diff
@@ -0,0 +1,114 @@
+        this.context = null;
+    }
+
+    rethrowIfCritical(err) {
```
This can be a freestanding function, it doesn't need an instance of the camera.

RReverser commented on ui/camera.js:
```diff
@@ -0,0 +1,114 @@
+    }
+
+    async connect() {
+        this.Module = await ModulePromise;
```
Maybe lazily assign `ModulePromise` here? On one hand, this approach is faster, but on another if someone just imported the library but never connects to a camera, this can save some bandwidth / startup time.

icheered replied:
I'm not 100% sure if I implemented it correctly, but this is what I could come up with.

RReverser commented on ui/camera.js:
```diff
@@ -33,32 +33,34 @@ async function initializeModule() {
-        if (err.constructor !== Error) {
-            throw err;
-        }
+        this.ModulePromise = null;
```
Module is shared between all instances, so this can be a global let variable instead.

icheered replied:
I'm not exactly sure what you mean with this (and if I already fixed it in later commits). Do you mean there is a single Module object for multiple Camera classes and therefore should be separated from the Camera class? I'm resolving for now, if it is a dealbreaker I'm happy to fix but otherwise I'd say leave it be :)

RReverser commented on README.md:
```diff
@@ -1,9 +1,100 @@
+
+async function connectCamera() {
+    // @ts-ignore
+    await navigator.usb.requestDevice({
```
I don't remember, but this is the generic selector for DSLR cameras, right? If so, this call should probably also live in the module like `showCameraPicker` or something so that users don't have to remember or copy-paste those values.

RReverser commented on package.json:
```diff
@@ -1,8 +1,35 @@
+    "libusb",
+    "webassembly"
+  ],
+  "author": "Ingvar Stepanyan - RReverser - me@rreverser.com",
```
Check out either JSON or shorthand npm syntax for people fields: https://docs.npmjs.com/cli/v9/configuring-npm/package-json#people-fields-author-contributors

I usually just put "Ingvar Stepanyan <me@rreverser.com>" which is valid shorthand.

Also add `contributors` field with yourself in the list too, after all, you're authoring the package :)

icheered commented:
I moved your demo into a separate `examples` folder as I intend to add a Sveltekit example in there as well once the package is on NPM :)

icheered commented:
@RReverser I have made some more minor changes and I'm almost done with the Svelte demo but I want to make a separate PR for that. I also confirmed that video is working as expected with the new implementation!

I think its ready for publishing :)

RReverser commented on README.md:
```diff
@@ -1,9 +1,106 @@
+
+async function capturePreviewAsBlob() {
+  const blob = await camera.capturePreviewAsBlob();
+  console.log("Blob:", blob);
```
Let's show a bit more realistic example here. Logged blob or file is not very useful, we could instead show e.g. creating `img` element from it.

icheered replied:
Added

RReverser commented on README.md:
```diff
@@ -1,9 +1,106 @@
+
+### Error: Not found: /node_modules/.vite/deps/libapi.wasm
+
+Vite tries to optimize the dependencies by default. This causes the WebAssembly module to be moved to a different location. To prevent this, you need to exclude the web-gphoto2 package from the optimization.
```
Not sure how I feel about including framework-specific workarounds tbh, both because of noise for non-Vite users in README and because we need to maintain them when/if this changes... is there a link to docs or stackoverflow or something could point to here instead?

icheered replied:
I completely understand your concern but this is not a widely documented issue and it took me a while to figure out why I was getting weird errors during development. I found the information in this issue: https://github.com/vitejs/vite/issues/8497 which is a duplicate of this issue: https://github.com/vitejs/vite/issues/8427 but that issue is still open and is not one that is found easily when looking to solve the issue.

The way I see it: This short message saves some people a lot of time and only takes little time (just reading the title is enough) for those that don't need it. 

What do you think of moving the Common Issues section all the way to the bottom? That way people looking for any other information won't even see it.

RReverser replied:
Hmm I see. Maybe just hide it in collapsed `<details>` for now, I hope we'll at least be able to remove it in the future. https://gist.github.com/scmx/eca72d44afee0113ceb0349dd54a84a2

icheered replied:
I moved the common issues section all the way to the bottom (just before 'See also' and 'License')

RReverser commented on examples/preact/ops.js:
```diff
@@ -25,7 +25,8 @@ const ModulePromise = initModule();
   // WebAssembly.RuntimeError, TypeError, etc. - is treated as critical here.
-  if (err.constructor !== Error) {
+  if (err.constructor !== Error)
+  {
```
Nit: what happened to formatting here and below with braces moving onto new lines? Can you fix manually or run them through Prettier please?

RReverser replied:
Also looks like those functions are now duplicated with camera class? Can we reuse them instead? (by updating the demo)

icheered replied:
I pulled the files through prettier.

We _can_ update the preact demo (and probably should), but since I'm very much not experienced with React ~and thoroughly dislike working with it~ I kind of wanted to leave that up to you ;)

I made the changes to the repo in such a way that the demo wouldn't be broken precisely for this reason

RReverser commented on package.json:
```diff
@@ -1,8 +1,47 @@
+      "url": "https://icheered.com"
+    }
+  ],
+  "license": "LGPL-2.1-or-later",
```
Hm why did you change the license string? Let's keep it as it was I think.

icheered replied:
During testing I got a warning that LGPLv2 is not a valid license string. 
https://docs.npmjs.com/cli/v9/configuring-npm/package-json#license
It should be one of the following:
https://spdx.org/licenses/

RReverser replied:
Huh okay.

RReverser commented on src/camera.js:
```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright 2021 Google LLC
```
This should have current year since it was just created.

RReverser commented on src/camera.js:
```diff
@@ -0,0 +1,131 @@
+let initModule;
+
+async function initializeModule() {
+  if (typeof window !== "undefined") {
```
What will happen if you don't do manual check? I assume it'd fail anyway? I'd rather not use manual condition, especially since we might still add Node.js support via [node-usb](https://www.npmjs.com/package/usb) in the future (I tried it once, it was a bit buggy back then, but might be better today).

icheered replied:
I just tested and adding the ENVIRONMENT flags during build stage removes all the server-side code so this is no longer needed.

RReverser commented on src/libapi.mjs.d.ts:
```diff
@@ -55,6 +58,27 @@ declare class Context {
   isDeleted(): boolean;
 }
 
+export declare class Camera {
```
Camera is not part of libapi.mjs though. This `.d.ts` is only to describe types in the Emscripten-generated `libapi.mjs`.

For `camera.js` you can describe types inline in JSDoc comments instead so that they're colocated with code, and also so that TypeScript is not confused and doesn't think that `Camera` lives in `libapi.mjs` module.

icheered replied:
Question because I'm not used to JSDoc: Can you still get type hints in typescript? I'm testing right now after adding the JSDoc syntax to the Camera class in `camera.js` and removing the definition from `libapi.d.ts` but type hints no longer work.

I used to do:
```ts
import Camera from 'web-gphoto2-npm-test';
import type { Camera as CameraClass } from 'web-gphoto2-npm-test';

export const camera: CameraClass = new Camera();
```
![image](https://github.com/GoogleChromeLabs/web-gphoto2/assets/37150873/1994fa76-86df-4b7c-b2ef-3437feefc575)


But right now it tells me that the package does not export a type called Camera, and removing the type definition all together makes using the class a bit more difficult because the functions are not showing up as type hints:

![image](https://github.com/GoogleChromeLabs/web-gphoto2/assets/37150873/d23b9169-8f29-41e6-ac89-7db48360f89b)

RReverser replied:
Yes, they're supposed to work, both hints and actual typechecking. If you can't get it to work I might take a look later.

RReverser commented:
Some final changes needed, but otherwise this is shaping up nicely, thanks :)

icheered commented:
@RReverser Implemented your feedback!

RReverser commented on package.json:
```diff
@@ -1,8 +1,47 @@
+  "description": "WebAssembly implementation of gphoto2 and libusb to control DSLR cameras over USB on the Web",
+  "type": "module",
+  "main": "src/camera.js",
+  "types": "src/libapi.mjs.d.ts",
```
Ah I suspect it might be because of this field that specifies types for the whole package. Try removing it?

icheered replied:
Hmm it seems that NPM packages pretty much require the 'types' field. If I remove it I get the following error when importing the package:

```
Could not find a declaration file for module 'web-gphoto2-npm-test'. '/home/tjbakker/Documents/dev/vscode/webdev/EditExpress/node_modules/web-gphoto2-npm-test/src/camera.js' implicitly has an 'any' type.
  If the 'web-gphoto2-npm-test' package actually exposes this module, try adding a new declaration (.d.ts) file containing `declare module 'web-gphoto2-npm-test';`ts(7016)
 ```

RReverser replied:
Ah you probably need to follow this instead. It will generate `.d.ts` for npm package from the JS file: https://www.typescriptlang.org/docs/handbook/declaration-files/dts-from-js.html

icheered replied:
I tried throwing a cheeky 
```ts/**
 * @typedef {import('./camera.js').Camera} Camera
 */```
 
Into the libapi.mjs.d.ts file but that didn't work. I could construct a `types.d.ts` file containing the previous class definition, and then also import the type definitions from `libapi.d.ts` so that we have a single file with all type definitions? I'm pretty sure that would work but I'm not sure if that is the normal way to do things.

RReverser replied:
Yeah I guess just creating a separate `types.d.ts` is fine for now. You should be able to use `import('./libapi.mjs')` and such there.

icheered replied:
Fixed it

RReverser commented:
Hm my follow-up comment disappeared for some reason, but please fix `ModulePromise` as requested earlier - it needs to be a singleton stored outside of `Camera` class, yes, to avoid recompiling / re-instantiating same Wasm module on each camera instance.

Browser caching makes it a bit less of a problem but it's still not great to recreate the whole "world" for camera instantiation.

icheered commented:
Alright I added the `types.d.ts` file with a class definition for Camera and removed the jsdoc syntax from `camera.js`. Now the types are automatically inferred 🎉. I made Camera a named export instead of a default export since this is needed for the automatic inference. I also moved  ModulePromise outside the Camera class.

RReverser commented on src/camera.js:
```diff
@@ -0,0 +1,105 @@
+    this.context = null;
+  }
+
+  async showCameraPicker() {
```
Sorry, one more nit: this should be either standalone function or perhaps better yet a static method, as it's not tied to the instance. For example:

```suggestion
  static async showPicker() {
```

so user can call it as `await Camera.showPicker()`.

RReverser commented on src/camera.js:
```diff
@@ -0,0 +1,105 @@
+  }
+}
+
+export { Camera };
```
Very minor nit, but you could inline this at the declaration instead so it's just `export class Camera { ... }`.

RReverser commented on src/libapi.mjs.d.ts:
```diff
@@ -41,6 +41,9 @@ declare interface SupportedOps {
 
 declare class Context {
+  context(): Promise<any> {
+    throw new Error('Method not implemented.');
```
Huh, does this actually compile? I thought `.d.ts` are not allowed to contain any executable code, only types. When would it even be reached?

icheered replied:
I hadn't noticed that, I genuinely don't know when that slipped in there

RReverser commented:
Thanks, I think this is looking good! I'll look into updating the official demo (aka what used to be the main app).

icheered commented:
Awesome, thanks @RReverser! Could you also publish the package to NPM using `npm publish`? That way I can immediately use it for a project I'm working on :)

RReverser commented:
I'm trying it out locally now and there are some more minor updates I need to do before publishing, but hopefully will do those soon.

RReverser commented on src/camera.js:
```diff
@@ -0,0 +1,103 @@
+  async setConfigValue(name, value) {
+    const uiTimeout = new Promise((resolve) => setTimeout(resolve, 800));
+    const setResult = this.schedule((context) =>
+      context.setConfigValue(name, value)
```
Also only just noticed that this got broken when migrating. Previously `uiTimeout` would only start when the scheduled operation is reached (inside the callback), but now it's assigned outside the callback, which makes it start much earlier so it can timeout too early as well.

RReverser commented:
Hm I don't see anything relevant that was changed by this PR, but looks like `./build.sh` started to fail, both locally and on CI:

```
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... configure: error: in `/src/deps/libtool':
configure: error: cannot run C compiled programs.
If you meant to cross compile, use `--host'.
See `config.log' for more details
make: *** [Makefile:40: deps/libtool/Makefile] Error 1
emmake: error: 'make -j8' failed (returned 2)
```

The logs are expired but it clearly still worked in March when I made last update: https://github.com/GoogleChromeLabs/web-gphoto2/actions/runs/4338585581/job/11777833871

So weird...

RReverser commented:
Ah. Here are extended logs from `config.log`:

```
configure:4044: /emsdk/upstream/emscripten/emcc -o conftest -Os -flto  -L/sysroot/lib -s DYNAMIC_EXECUTION=0 -s AUTO_JS_LIBRARIES=0 -s AUTO_NATIVE_LIBRARIES=0 -s ENVIRONMENT=web,worker -Os -flto conftest.c  >&5
configure:4048: $? = 0
configure:4055: ./conftest
internal/process/esm_loader.js:74
    internalBinding('errors').triggerUncaughtException(
                              ^

TypeError [ERR_UNKNOWN_FILE_EXTENSION]: Unknown file extension "" for /src/deps/libtool/conftest
    at new NodeError (internal/errors.js:322:7)
    at Loader.defaultGetFormat [as _getFormat] (internal/modules/esm/get_format.js:71:15)
    at Loader.getFormat (internal/modules/esm/loader.js:105:42)
    at Loader.getModuleJob (internal/modules/esm/loader.js:243:31)
    at async Loader.import (internal/modules/esm/loader.js:177:17)
    at async Object.loadESM (internal/process/esm_loader.js:68:5) {
  code: 'ERR_UNKNOWN_FILE_EXTENSION'
}
configure:4059: $? = 1
configure:4066: error: in `/src/deps/libtool':
configure:4068: error: cannot run C compiled programs.
If you meant to cross compile, use `--host'.
See `config.log' for more details
```

I think setting `"type": "module"` in package.json in this PR broke it as it now affects all files produced inside the project, even those generated by autotools.

RReverser commented:
Ok fixed it by switching to `"exports"` field in package.json. It's somewhat newer, but I hope all modern bundlers will pick it up correctly.

RReverser commented:
Heh I reverted version to 0.1.0 since I'm pretty sure this will require some more changes before considered 1.0, but then getting 

npm ERR! 400 Bad Request - PUT https://registry.npmjs.org/web-gphoto2 - Cannot publish over previously published version "0.1.0".

I guess even if you delete package, the version restriction remains.

Oh well, I'll update to 0.2.0 or something.

RReverser commented:
Ah... you published 0.2.0 before as well. Sigh :( I'll keep trying.

FWIW for the future if you're just testing you can use dependency via `file:../path/to/package` instead of publishing to npm. Or publish under a throwaway name maybe.

RReverser commented:
Reached 0.4.0, looks like that finally worked.

icheered commented:
> I guess even if you delete package, the version restriction remains.

I did not expect that... I did start using a throwaway name but as this is the first time I worked on publishing through NPM I've been learning a lot about the process

> Reached 0.4.0, looks like that finally worked.

Awesome! Luckily there are infinite sub-versions available before publishing a major release ;)

RReverser commented:
Heh yeah. I just published 0.4.1 too as apparently needed few more fields for it to resolve on unpkg. Seems to work now, fingers crossed for the demo deploy.

icheered commented:
I'll test it later tonight and will try to have a svelte demo PR ready soon. Thanks for you help on the project, I'm happy I could contribute :)

RReverser commented:
Whew the demo seems to work.
